docs/tests.md

```diff
@@ -0,0 +1,345 @@
+to a specific directory within the `tests/` directory:
+
+1. **Generating Expected Outputs:**
+
+   - The `references/` directory stores **generator** functions for producing
+     the expected outputs of the estimators.
+   - Each estimator has a dedicated subdirectory under `references/`, e.g.,
+     `<estimator_name>Estimator/`.
+   - Inside the estimator's subdirectory:
+     - `legacy_implementations/` (optional): Contains previous implementations
+       of the estimator.
+     - `gen_<estimator_name>.py`: Houses the generator functions.
+
+   Generator functions must adhere to the following conventions: - They start
+   with the prefix `gen_`. - They have a single parameter named `inputs`, which
+   can be either a single test case (tuple) or a tuple containing information to
+   generate multiple test cases. - They return a list of tuples, where each
+   tuple is of the form `(input, expected_output)`.
+
+1. **Defining Test Inputs:**
+
+   - The `tests/references/generate_references.py` file holds the
+     `REFERENCES_INPUTS` dictionary.
+   - This dictionary mirrors the file structure of the `references/` directory
+     and specifies the inputs for the generator functions.
+   - To add a new estimator and its inputs, follow this structure within the
+     dictionary:
+
+   ```python
+   REFERENCES_INPUTS = {
+       "<estimator_name>Estimator": {
+           "gen_<estimator_name>": {
+               "gen_<estimator_name>_<function_name_1>": {
+                   "inputs": [(256, 128, 64, 251), (961, 771, 48, 31)]
+               },
+               # Add more functions as needed...
+           },
+       },
+       # Add more estimators as needed...
+   }
+   ```
+
+   After defining the inputs, run the following command from the library root to
+   generate the reference values:
+
+   ```bash
+   make docker-generate-tests-references
+   ```
+
+   **Warning:** This command creates a Docker container with SageMath installed,
+   which may take some time to run.
+
+   The generated inputs and their corresponding outputs will be serialized in
+   `tests/references/reference_values.yaml`.
+
+1. **Implementing Tests:**
+
+   - The `tests/validations/` directory contains the test implementation files.
+
+   - Each estimator has a corresponding file named `test_<estimator_name>.py`.
+
+   - For each generator function (`gen_<estimator_name>_<function_name>`) in
+     `references/`, there must be a corresponding test function named
+     `test_<estimator_name>_<function_name>` in the validation file.
+
+   - Each test function:
+
+     - Takes a single parameter, `test_data`, automatically provided by _pytest_
+       and containing the input/output pairs from the `reference_values.yaml`
+       file.
+     - The boilerplate code is expected to have the following structure:
+
+     ```python
+     def test_<my_estimator>_<my_function>(test_data):
+         """
+         Test for the ... problem with an error tolerance of epsilon.
```

Comment:
Docstrings should avoid hardcoded values. This prevents contributors from having to manually update the information in both the docstring and the code. Since this value can vary between tests, referencing the `epsilon` variable defined on the next line is a more maintainable approach.